pkg/querier/queryrange/split_by_range.go

```diff
@@ -0,0 +1,120 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/tenant"
+	util_log "github.com/grafana/loki/pkg/util/log"
+	"github.com/grafana/loki/pkg/util/marshal"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type splitByRange struct {
+	logger log.Logger
+	next   queryrangebase.Handler
+	limits Limits
+
+	ng *logql.DownstreamEngine
+}
+
+// SplitByRangeMiddleware creates a new Middleware that splits log requests by the range interval.
+func SplitByRangeMiddleware(logger log.Logger, limits Limits, metrics *logql.ShardingMetrics) queryrangebase.Middleware {
```

Comment:
Can you explain why we need a specific middleware for this? I figured we could handle this in AST-mapping and use the existing sharded middleware to dispatch the resulting queries.

Reply:
The idea was to have two separate middlewares each one with its own responsibility (sharding or splitting). This way we can have a tripperware that does just sharding whereas another one that does just splitting. This would also facilitate the process of debugging/testing each middleware separately.
But the final idea is to have a tripperware that has both the sharding and the splitting middlewares 👍